Abstract facebook _extract_metadata method out from inside '_extract_from_url'

pjrobertson opened the pull request:
<!--

Explanation of your *pull request* in arbitrary form goes here. Please **make sure the description explains the purpose and effect** of your *pull request* and is worded well enough to be understood. Provide as much **context and examples** as possible

-->

I may be pushing my luck with this change, but this abstracts the very useful `_extract_metadata` method out from inside the `_extract_from_url` method. This is similar to what was merged for the bluesky extractor last week [here](https://github.com/yt-dlp/yt-dlp/pull/12194).

This allows external 3rd party apps to call into this, of course with the caveats that this is a private API and it may change at any point.


### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)

</details>

bashonly commented:
IMO if we're gonna have this huge diff in the FB extractor, we might as well doing something actually useful for yt-dlp too and move all of those icky nested functions out into methods (where it is possible)

bashonly commented:
Yeah looks fine. I would also move out `yield_all_relay_data` and `extract_relay_prefetched_data` and then call it good

pjrobertson commented:
Alright, I've brought those two (and one more) out. Hopefully this is good to go now!

pjrobertson commented:
Any chance this can get merged this week? Would be great to be able to use this in an upcoming release of the auto-archiver.

Much appreciated!

pjrobertson commented:
OK, I have done some more cleaning up:

1. Removed the video_id from the signature 
2. Investigated the tahoe player, it comes back to this issue: https://github.com/ytdl-org/youtube-dl/issues/15441 which was 'fixed' in https://github.com/yt-dlp/yt-dlp/commit/9d082e7cb81c63dc5c2e616cd7ca7237a5e0d642 . I have added that problematic URL (https://www.facebook.com/WatchESLOne/videos/297860117405429/) to the tests, but it still fails now - meaning the Tahoe code didn't work (tahoe URL just returned 'not found' for all videos I tried), so I have removed the tahoe code.

You're right, this extractor is horrible!

pjrobertson commented:
Sorry, I will not do it again. I thought it would be OK since it was just overwriting one commit from 2 minutes earlier, but I've  learnt my lesson 👍
